slides.typ

```diff
@@ -1,23 +1,222 @@
+        if match-until != none {
+            let parsed = int(match-until.captures.first())
+            // assert(parsed > 0, "parsed idx is non-positive")
+            ( until: parsed )
+        } else if match-beginning != none {
+            let parsed = int(match-beginning.captures.first())
+            // assert(parsed > 0, "parsed idx is non-positive")
+            ( beginning: parsed )
+        } else if match-range != none {
+            let parsed-first = int(match-range.captures.first())
+            let parsed-last = int(match-range.captures.last())
+            // assert(parsed-first > 0, "parsed idx is non-positive")
+            // assert(parsed-last > 0, "parsed idx is non-positive")
+            ( beginning: parsed-first, until: parsed-last )
+        } else if match-single != none {
+            let parsed = int(match-single.captures.first())
+            // assert(parsed > 0, "parsed idx is non-positive")
+            parsed
+        } else {
+            panic("failed to parse visible slide idx:" + part)
+        }
+    }
+    parts.map(parse-part)
+}
+
+#let _check-visible(idx, visible-subslides) = {
+    if type(visible-subslides) == "integer" {
+        idx == visible-subslides
+    } else if type(visible-subslides) == "array" {
+        visible-subslides.any(s => _check-visible(idx, s))
+    } else if type(visible-subslides) == "string" {
+        let parts = _parse-subslide-indices(visible-subslides)
+        _check-visible(idx, parts)
+    } else if type(visible-subslides) == "dictionary" {
+        let lower-okay = if "beginning" in visible-subslides {
+            visible-subslides.beginning <= idx
+        } else {
+            true
+        }
+
+        let upper-okay = if "until" in visible-subslides {
+            visible-subslides.until >= idx
+        } else {
+            true
+        }
+
+        lower-okay and upper-okay
+    } else {
+        panic("you may only provide a single integer, an array of integers, or a string")
+    }
+}
+
+#let _last-required-subslide(visible-subslides) = {
+    if type(visible-subslides) == "integer" {
+        visible-subslides
+    } else if type(visible-subslides) == "array" {
+        calc.max(..visible-subslides.map(s => _last-required-subslide(s)))
+    } else if type(visible-subslides) == "string" {
+        let parts = _parse-subslide-indices(visible-subslides)
+        _last-required-subslide(parts)
+    } else if type(visible-subslides) == "dictionary" {
+        let last = 0
+        if "beginning" in visible-subslides {
+            last = calc.max(last, visible-subslides.beginning)
+        }
+        if "until" in visible-subslides {
+            last = calc.max(last, visible-subslides.until)
+        }
+        last
+    } else {
+        panic("you may only provide a single integer, an array of integers, or a string")
+    }
+}
+
+#let _conditional-display(visible-subslides, reserve-space, mode, body) = {
```

Comment:
I'm not sure I understand your point. With your suggestions, you can call `#only` with a nonsensical mode. But the current implementation doesn't allow that. Users can only do something inconsistent when calling `_conditional-display` directly.

Reply:
Ohh you mean your proposed change introduces this potential inconsistency. Yes, that is the reason why I use the extra `_conditional-display`, otherwise I would make `uncover` call `only`.